src/database_handler.cc

```diff
@@ -0,0 +1,256 @@
+    std::string result;
+    std::string::size_type i;
+    for (i = 0; i < str.size(); ++i) {
+        if (str[i] == '+') {
+            result += ' ';
+        }
+        else if (str[i] == '%' && str.size() > i+2) {
+            const unsigned char ch1 = FromHex(str[i+1]);
+            const unsigned char ch2 = FromHex(str[i+2]);
+            const unsigned char ch = (ch1 << 4) | ch2;
+            result += ch;
+            i += 2;
+        }
+        else {
+            result += str[i];
+        }
+    }
+    return result;
+}
+
+const std::string DatabaseHandler::ExtractQuery(const std::string& uri) {
+    // Example URI: /database?query=select+name+from+actors
+    std::string param = "query=";
+
+    // Query starts after "query="
+    std::size_t query_start = uri.find(param);
+    if (query_start == std::string::npos) {
+        // No query parameter
+        return "";
+    }
+
+    // Decode the URI
+    return URLDecode(uri.substr(query_start + param.length()));
+}
+
+const std::string DatabaseHandler::GetJSONResults(sql::Statement *stmt) {
+    // Get rows
+    sql::ResultSet *results = stmt->getResultSet();
+
+    // Get column information
+    sql::ResultSetMetaData *col_meta = results->getMetaData();
+
+    /* Build JSON Response
+     * Format:
+     * {
+     *     "cols": ["col1", "col2", ... , "coln"],
+     *     "data": {
+     *          "rowID_1": ["attr1", "attr2", ... , "attrn"],
+     *          "rowID_2": ["attr1", "attr2", ... , "attrn"],
+     *          ...
+     *          "rowID_n": ["attr1", "attr2", ... , "attrn"]
+     *     }
+     * }
+     */
+    std::string result_string = "{\n";
+
+    // Print column names
+    result_string += "\"cols\": [";
+    for (size_t i = 1; i <= col_meta->getColumnCount(); i++) {
+        std::string col_name = col_meta->getColumnName(i);
+        result_string += "\"" + col_name + "\"";
+
+        // Append comma to end if not last element
+        if (i != col_meta->getColumnCount()) {
+            result_string += ",";
+        }
+    }
+    result_string += "],\n\"data\": {\n";
+
+    // Print row values
+    while (results->next()) {
+        int row_id = results->getRow();
+        result_string += "\"" + std::to_string(row_id) + "\": [";
+
+        // Columns are 1-indexed
+        for (size_t i = 1; i <= col_meta->getColumnCount(); i++) {
+            std::string value = results->getString(i);
+            result_string += "\"" + value + "\"";
+
+            // Append comma to end if not last element
+            if (i != col_meta->getColumnCount()) {
+                result_string += ",";
+            }
+        }
+        result_string += "],\n";
+    }
+
+    // Remove trailing comma from last data element (if exists)
+    if (results->rowsCount() > 0) {
+        result_string = result_string.substr(0, result_string.length()-2);
+    }
+    result_string += "}\n}";
+
+    // Free resultSet and metadata pointer
```

Comment:
Oh oops that comment is out of date. I tried to free the metadata pointer, but I think we're not supposed to handle that, because the destructor is protected.